File: pcc/py_runtime/src/py_os_env.c
/* pcc/py_runtime/src/py_os_env.c
 *
 * Environment-variable helpers split out of py_os.c so the pcc-
 * Python port (py_os_env.py) can replace just these three symbols
 * in the libpy_runtime_pcc_py.a archive without touching the path
 * helpers that still live in py_os_path.c.
 */

#include "py_internal.h"

#include <stdlib.h>
#include <string.h>

static PyObject *coerce_path_str(PyObject *o, PyObject **owned) {
    *owned = NULL;
    if (o == NULL) return NULL;
    if (py_type_of(o) == PY_TYPE_STR) return o;
    *owned = py_obj_str(o);
    return *owned;
}

PyObject *py_os_getenv(PyObject *key, PyObject *default_value) {
    PyObject *owned = NULL;
    PyObject *item = coerce_path_str(key, &owned);
    if (item == NULL) {
        py_decref(owned);
        return default_value;
    }
    const char *name = py_str_utf8(item);
    if (name == NULL) {
        py_decref(owned);
        return default_value;
    }
    const char *raw = getenv(name);
    py_decref(owned);
    if (raw == NULL) return default_value;
    return py_str_new(raw, (int64_t)strlen(raw));
}

PyObject *py_os_putenv(PyObject *key, PyObject *value) {
    PyObject *owned_key = NULL;
    PyObject *owned_value = NULL;
    PyObject *key_obj = coerce_path_str(key, &owned_key);
    PyObject *value_obj = coerce_path_str(value, &owned_value);
    if (key_obj == NULL || value_obj == NULL) {
        py_decref(owned_key);
        py_decref(owned_value);
        return py_None;
    }
    const char *raw_key = py_str_utf8(key_obj);
    const char *raw_value = py_str_utf8(value_obj);
    if (raw_key != NULL && raw_value != NULL) {
        (void)setenv(raw_key, raw_value, 1);
    }
    py_decref(owned_key);
    py_decref(owned_value);
    return py_None;
}

/* os.environ[key]: CPython mapping semantics — the key must be a str
 * (TypeError otherwise, like CPython's encodekey()) and a missing
 * variable raises KeyError carrying the key. py_os_getenv stays
 * non-raising for os.getenv() / os.environ.get(). */
PyObject *py_os_environ_getitem(PyObject *key) {
    if (key == NULL || py_type_of(key) != PY_TYPE_STR) {
        py_raise_owned(py_exc_new(PY_EXC_TYPEERROR, "str expected"));
        return NULL;
    }
    const char *name = py_str_utf8(key);
    if (name == NULL) {
        py_raise_owned(py_exc_new_with_value(PY_EXC_KEYERROR, key));
        return NULL;
    }
    const char *raw = getenv(name);
    if (raw == NULL) {
        py_raise_owned(py_exc_new_with_value(PY_EXC_KEYERROR, key));
        return NULL;
    }
    return py_str_new(raw, (int64_t)strlen(raw));
}

int32_t py_os_environ_contains(PyObject *key) {
    if (key == NULL || py_type_of(key) != PY_TYPE_STR) {
        py_raise_owned(py_exc_new(PY_EXC_TYPEERROR, "str expected"));
        return -1;
    }
    const char *name = py_str_utf8(key);
    if (name == NULL) {
        py_raise_owned(py_exc_new(PY_EXC_TYPEERROR, "str expected"));
        return -1;
    }
    return getenv(name) != NULL ? 1 : 0;
}

/* os.environ[key] = value: CPython mapping semantics — both key and
 * value must be str (TypeError otherwise); the store is visible to the
 * process environment (setenv), matching CPython's putenv-backed
 * __setitem__. py_os_putenv stays coercing/non-raising. */
PyObject *py_os_environ_setitem(PyObject *key, PyObject *value) {
    if (key == NULL || py_type_of(key) != PY_TYPE_STR) {
        py_raise_owned(py_exc_new(PY_EXC_TYPEERROR, "str expected"));
        return NULL;
    }
    if (value == NULL || py_type_of(value) != PY_TYPE_STR) {
        py_raise_owned(py_exc_new(PY_EXC_TYPEERROR, "str expected"));
        return NULL;
    }
    const char *raw_key = py_str_utf8(key);
    const char *raw_value = py_str_utf8(value);
    if (raw_key != NULL && raw_value != NULL) {
        (void)setenv(raw_key, raw_value, 1);
    }
    return py_None;
}

PyObject *py_os_unsetenv(PyObject *key) {
    PyObject *owned = NULL;
    PyObject *item = coerce_path_str(key, &owned);
    if (item == NULL) {
        py_decref(owned);
        return py_None;
    }
    const char *raw = py_str_utf8(item);
    if (raw != NULL) {
        (void)unsetenv(raw);
    }
    py_decref(owned);
    return py_None;
}

File: pcc/py_runtime/py/py_os_env.py
"""Phase 4c.9: pcc-Python port of py_os_env.c.

Environment-variable helpers (getenv / putenv / unsetenv). The path
helpers stay in py_os_path.c for now because they need mutable byte-
buffer growth that pcc-Python cannot yet express cleanly.

Coercion: keys/values may be any py object; non-str gets routed
through py_obj_str() to realize a PyStrObject before reading its
UTF-8 bytes.
"""

from pcc.extern import extern, c_abi_export, c_ptr, c_int64, c_void
from pcc.unsafe import (
    cstr,
    getenv,
    global_load_ptr,
    is_tagged_int,
    load_i32,
    null,
    ptr_is_null,
    setenv,
    strlen,
    unsetenv,
)

py_decref = extern("py_decref", (c_ptr,), c_void)
py_str_new = extern("py_str_new", (c_ptr, c_int64), c_ptr)
py_str_utf8 = extern("py_str_utf8", (c_ptr,), c_ptr)
py_obj_str = extern("py_obj_str", (c_ptr,), c_ptr)
py_exc_new = extern("py_exc_new", (c_int64, c_ptr), c_ptr)
py_exc_new_with_value = extern("py_exc_new_with_value", (c_int64, c_ptr), c_ptr)
py_raise_owned = extern("py_raise_owned", (c_ptr,), c_void)


# PY_TYPE_STR (=4) is inlined at every use site because pcc-Python
# initializes module-level integers in the auto-generated main(),
# which the Makefile strips for library .o builds.


def _type_of(obj) -> int:
    if is_tagged_int(obj):
        return 2  # PY_TYPE_INT
    return load_i32(obj, 8)


def _coerce_to_str(o):
    # Returns (item, owned) — item is the PyStrObject to read, owned
    # is the temporary we must decref after, or NULL.
    if ptr_is_null(o):
        return null(), null()
    if _type_of(o) == 4:  # PY_TYPE_STR
        return o, null()
    new_s = py_obj_str(o)
    return new_s, new_s


@c_abi_export("py_os_getenv")
def py_os_getenv(key, default_value):
    item, owned = _coerce_to_str(key)
    if ptr_is_null(item):
        if not ptr_is_null(owned):
            py_decref(owned)
        return default_value
    name = py_str_utf8(item)
    if ptr_is_null(name):
        if not ptr_is_null(owned):
            py_decref(owned)
        return default_value
    raw = getenv(name)
    if not ptr_is_null(owned):
        py_decref(owned)
    if ptr_is_null(raw):
        return default_value
    n: int = strlen(raw)
    return py_str_new(raw, n)


@c_abi_export("py_os_environ_contains")
def py_os_environ_contains(key) -> int:
    if ptr_is_null(key) != 0:
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return -1
    if _type_of(key) != 4:  # PY_TYPE_STR
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return -1
    name = py_str_utf8(key)
    if ptr_is_null(name) != 0:
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return -1
    if ptr_is_null(getenv(name)) != 0:
        return 0
    return 1


@c_abi_export("py_os_putenv")
def py_os_putenv(key, value):
    k_item, k_owned = _coerce_to_str(key)
    v_item, v_owned = _coerce_to_str(value)
    if ptr_is_null(k_item):
        if not ptr_is_null(k_owned):
            py_decref(k_owned)
        if not ptr_is_null(v_owned):
            py_decref(v_owned)
        return global_load_ptr("py_None")
    if ptr_is_null(v_item):
        if not ptr_is_null(k_owned):
            py_decref(k_owned)
        if not ptr_is_null(v_owned):
            py_decref(v_owned)
        return global_load_ptr("py_None")
    k_raw = py_str_utf8(k_item)
    v_raw = py_str_utf8(v_item)
    if not ptr_is_null(k_raw):
        if not ptr_is_null(v_raw):
            setenv(k_raw, v_raw, 1)
    if not ptr_is_null(k_owned):
        py_decref(k_owned)
    if not ptr_is_null(v_owned):
        py_decref(v_owned)
    return global_load_ptr("py_None")


@c_abi_export("py_os_environ_getitem")
def py_os_environ_getitem(key):
    # os.environ[key]: CPython mapping semantics — the key must be a
    # str (TypeError otherwise, like CPython's encodekey()) and a
    # missing variable raises KeyError carrying the key. Mirrors
    # py_os_environ_getitem in py_os_env.c; py_os_getenv stays
    # non-raising for os.getenv() / os.environ.get().
    if ptr_is_null(key) != 0:
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    if _type_of(key) != 4:  # PY_TYPE_STR
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    name = py_str_utf8(key)
    if ptr_is_null(name) != 0:
        py_raise_owned(py_exc_new_with_value(4, key))  # PY_EXC_KEYERROR
        return null()
    raw = getenv(name)
    if ptr_is_null(raw) != 0:
        py_raise_owned(py_exc_new_with_value(4, key))  # PY_EXC_KEYERROR
        return null()
    n: int = strlen(raw)
    return py_str_new(raw, n)


@c_abi_export("py_os_environ_setitem")
def py_os_environ_setitem(key, value):
    # os.environ[key] = value: CPython mapping semantics — both key
    # and value must be str (TypeError otherwise); the store is
    # visible to the process environment (setenv), matching CPython's
    # putenv-backed __setitem__. Mirrors py_os_environ_setitem in
    # py_os_env.c; py_os_putenv stays coercing/non-raising.
    if ptr_is_null(key) != 0:
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    if _type_of(key) != 4:  # PY_TYPE_STR
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    if ptr_is_null(value) != 0:
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    if _type_of(value) != 4:  # PY_TYPE_STR
        py_raise_owned(py_exc_new(3, cstr("str expected")))  # PY_EXC_TYPEERROR
        return null()
    k_raw = py_str_utf8(key)
    v_raw = py_str_utf8(value)
    if ptr_is_null(k_raw) == 0:
        if ptr_is_null(v_raw) == 0:
            setenv(k_raw, v_raw, 1)
    return global_load_ptr("py_None")


@c_abi_export("py_os_unsetenv")
def py_os_unsetenv(key):
    item, owned = _coerce_to_str(key)
    if ptr_is_null(item):
        if not ptr_is_null(owned):
            py_decref(owned)
        return global_load_ptr("py_None")
    raw = py_str_utf8(item)
    if not ptr_is_null(raw):
        unsetenv(raw)
    if not ptr_is_null(owned):
        py_decref(owned)
    return global_load_ptr("py_None")
